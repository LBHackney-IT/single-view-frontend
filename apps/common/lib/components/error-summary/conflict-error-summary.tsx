import React from "react";

import locale from "../../locale";
import { SummaryList, SummaryListItem } from "../summary-list";
import { ErrorSummary, ErrorSummaryProps } from "./error-summary";

interface ConflictErrorSummaryProps extends Partial<ErrorSummaryProps> {
  id: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  updatedFields?: Record<string, any>;
  fieldLocale: Record<string, string>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  fieldTransforms?: Record<string, (value: any) => string>;
  footNote?: string;
}

const { changesNotSaved, anotherUserMadeEdit, youEntered, toSaveMakeEdit } =
  locale.components.conflictErrorSummary;

export const ConflictErrorSummary = ({
  updatedFields,
  fieldLocale,
  fieldTransforms,
  title = changesNotSaved,
  description = anotherUserMadeEdit,
  footNote = toSaveMakeEdit,
  ...props
}: ConflictErrorSummaryProps): JSX.Element => {
  const keys = Object.keys(updatedFields || {});
  return (
    <ErrorSummary
      className="mtfh-change-conflict"
      title={title}
      description={description}
      {...props}
    >
      {keys.length > 0 && updatedFields && (
        <>
          <p>{youEntered}</p>
          <SummaryList variant="inline">
            {Object.keys(updatedFields).map((key) => (
              <SummaryListItem key={key} title={`${fieldLocale[key] || key}:`}>
                {fieldTransforms && fieldTransforms[key]
                  ? fieldTransforms[key](updatedFields[key])
                  : `${updatedFields[key]}`}
              </SummaryListItem>
            ))}
          </SummaryList>
          {footNote && <p>{footNote}</p>}
        </>
      )}
    </ErrorSummary>
  );
};
