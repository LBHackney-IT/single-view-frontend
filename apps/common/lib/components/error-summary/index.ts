export * from "./error-summary";
export * from "./conflict-error-summary";
export * from "./form-error-summary";
export * from "./status-error-summary";
