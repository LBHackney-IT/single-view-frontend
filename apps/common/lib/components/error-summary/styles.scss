@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-error-summary/error-summary";
