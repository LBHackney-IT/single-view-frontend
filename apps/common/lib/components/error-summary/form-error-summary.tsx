import React from "react";

import locale from "../../locale";
import { ErrorSummary, ErrorSummaryProps } from "./error-summary";

interface FormErrorSummaryProps extends Partial<Omit<ErrorSummaryProps, "description">> {
  id: string;
  prefix: string;
  errors: Record<string, string>;
}

const { error } = locale.components.formErrorSummary;

export const FormErrorSummary = ({
  id,
  prefix,
  errors,
  title = error,
  ...props
}: FormErrorSummaryProps): JSX.Element => {
  return (
    <ErrorSummary id={id} title={title} {...props}>
      <ul className="govuk-list govuk-error-summary__list">
        {(Object.keys(errors) as Array<keyof typeof errors>)
          .filter((key) => errors[key])
          .map((key) => {
            return (
              <li key={key}>
                <a href={`#${prefix}-${key}`}>{errors[key]}</a>
              </li>
            );
          })}
      </ul>
    </ErrorSummary>
  );
};
