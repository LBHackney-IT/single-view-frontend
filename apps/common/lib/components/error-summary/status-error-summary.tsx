import React from "react";

import locale from "../../locale";
import { ErrorSummary, ErrorSummaryProps } from "./error-summary";

interface StautsErrorSummaryProps extends Partial<ErrorSummaryProps> {
  id: string;
  code: number;
}

const { statusTitle, statusDescription } = locale.components.statusErrorSummary;

export const StatusErrorSummary = ({
  id,
  code,
  title = statusTitle(code),
  description = statusDescription(code),
  ...props
}: StautsErrorSummaryProps): JSX.Element => {
  return <ErrorSummary id={id} title={title} description={description} {...props} />;
};
