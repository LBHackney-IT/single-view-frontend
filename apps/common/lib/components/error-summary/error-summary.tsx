import React, { forwardRef, useEffect, useRef } from "react";
import mergeRefs from "react-merge-refs";

import cn from "classnames";
import { ErrorSummary as ErrorSummaryJs } from "lbh-frontend";

import { widthOverrides } from "../../utils";

import type * as Polymorphic from "@radix-ui/react-polymorphic";
import "./styles.scss";

export interface ErrorSummaryProps {
  id: string;
  title: string;
  description?: string;
  reFocus?: number;
  override?: number;
}

export type ErrorSummaryComponent = Polymorphic.ForwardRefComponent<
  "div",
  ErrorSummaryProps
>;

export const ErrorSummary: ErrorSummaryComponent = forwardRef(function ErrorSummary(
  {
    as: ErrorSummaryComp = "div",
    id,
    title,
    description,
    className,
    children,
    reFocus,
    override,
    ...props
  },
  ref,
) {
  const localRef = useRef<HTMLElement>(null);

  useEffect(() => {
    /* istanbul ignore else */
    if (localRef.current) {
      // eslint-disable-next-line no-new
      new ErrorSummaryJs(localRef.current);
      localRef.current.scrollIntoView(true);
    }
  }, []);

  useEffect(() => {
    /* istanbul ignore else */
    if (localRef.current) {
      localRef.current.scrollIntoView(true);
    }
  }, [reFocus]);

  return (
    <ErrorSummaryComp
      ref={mergeRefs([localRef, ref])}
      className={cn(
        "govuk-error-summary",
        "lbh-error-summary",
        widthOverrides(override),
        className,
      )}
      aria-labelledby={id}
      role="alert"
      {...props}
    >
      <h2 className="govuk-error-summary__title" id={id}>
        {title}
      </h2>
      {description || children ? (
        <div className="govuk-error-summary__body">
          {description ? <p>{description}</p> : null}
          {children}
        </div>
      ) : null}
    </ErrorSummaryComp>
  );
});
