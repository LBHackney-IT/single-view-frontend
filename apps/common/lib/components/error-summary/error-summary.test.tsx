import React from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";
import { screen } from "@testing-library/react";

import { ErrorSummary } from "./error-summary";

test("it renders an alert", async () => {
  const { container } = render(<ErrorSummary id="error" title="Something went wrong" />);
  expect(screen.getByRole("alert")).toBeInTheDocument();
  await testA11y(container);
});

test("it shows a description without children", () => {
  render(
    <ErrorSummary id="error" title="Something went wrong" description="User error" />,
  );
  expect(screen.getByText("User error")).toBeInTheDocument();
});

test("it allows children", async () => {
  const { container } = render(
    <ErrorSummary id="error" title="Something went wrong">
      <p>Test</p>
      <div>Error</div>
    </ErrorSummary>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});
