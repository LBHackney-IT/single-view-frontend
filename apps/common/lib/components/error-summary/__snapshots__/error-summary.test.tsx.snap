// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it allows children 1`] = `
<div>
  <div
    aria-labelledby="error"
    class="govuk-error-summary lbh-error-summary"
    role="alert"
  >
    <h2
      class="govuk-error-summary__title"
      id="error"
    >
      Something went wrong
    </h2>
    <div
      class="govuk-error-summary__body"
    >
      <p>
        Test
      </p>
      <div>
        Error
      </div>
    </div>
  </div>
</div>
`;
